[threaded_pg] full rewrite of MultiThreadedTestCase to enable device_type tests

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #92198
* #92197
* __->__ #91650
* #91649

This PR did a full rewrite of MultiThreadedTestCase, to make it more
aligned with the MultiProcessTestCase, also changed how it do spawning
and testing, so that we could embed thread local states when running
tests.

This PR enables device_type tests to work with MultiThreadedTestCase

wanchaol committed:
[threaded_pg] refactor threaded pg to enable device_type tests

[ghstack-poisoned]

wanchaol committed:
Update on "[threaded_pg] refactor threaded pg to enable device_type tests"




[ghstack-poisoned]

wanchaol committed:
Update on "[threaded_pg] full rewrite of MultiThreadedTestCase to enable device_type tests"


This PR did a full rewrite of MultiThreadedTestCase, to make it more
aligned with the MultiProcessTestCase, also changed how it do spawning
and testing, so that we could embed thread local states when running
tests.

This PR enables device_type tests to work with MultiThreadedTestCase

[ghstack-poisoned]

XilunWu reviewed (approved):
lgtm. linter fix before landing.

wanchaol committed:
Update on "[threaded_pg] full rewrite of MultiThreadedTestCase to enable device_type tests"


This PR did a full rewrite of MultiThreadedTestCase, to make it more
aligned with the MultiProcessTestCase, also changed how it do spawning
and testing, so that we could embed thread local states when running
tests.

This PR enables device_type tests to work with MultiThreadedTestCase

[ghstack-poisoned]

wanchaol committed:
Update on "[threaded_pg] full rewrite of MultiThreadedTestCase to enable device_type tests"


This PR did a full rewrite of MultiThreadedTestCase, to make it more
aligned with the MultiProcessTestCase, also changed how it do spawning
and testing, so that we could embed thread local states when running
tests.

This PR enables device_type tests to work with MultiThreadedTestCase

[ghstack-poisoned]
